### PR TITLE
refactor: Introduce Pydantic

### PR DESCRIPTION
**Pull request by VRGhost**
## Overview

This branch adds `pydantic` configuration model and validation in dynamicio.

### CODE REVIEW

File: demo/tests/test_pipeline.py [hunk 1]
```diff
@@ -32,12 +32,24 @@ def test_dag_with_mock_sample_input_data(
...
+                pd.read_parquet(raw_config.get(source_key="STAGED_BAR").local.file_path)
+            )
+            pd.testing.assert_frame_equal(
+                expected_final_foo_df,
+                pd.read_parquet(processed_config.get(source_key="FINAL_FOO").local.file_path)
+            )
+            pd.testing.assert_frame_equal(
+                expected_final_bar_df,
+                pd.read_parquet(processed_config.get(source_key="FINAL_BAR").local.file_path)
+            )
```
Comment (Christos-Hadjinikolis) [inline demo/tests/test_pipeline.py:50]: Can you please configure pycharm to not mess with reformatting the line-length? Much readable if in one line

**Review by Christos-Hadjinikolis — COMMENTED** [review]
Big PR, we need a pairing session.

File: dynamicio/config/__init__.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+"""Dynamicio config file handling routines."""
+
+from . import pydantic
+from .io_config import IOConfig
```
[the remark this answers is not in the document]
Comment (VRGhost) [reply dynamicio/config/__init__.py:4]: I will have to disagree in the context of `__init__.py` files - assembling module namespace is much harder with absolute imports.

File: demo/tests/test_pipeline.py (hunk 1 above)
Comment (VRGhost) [reply demo/tests/test_pipeline.py:50]: This is `black` and it already aims for 120 character lines.

File: dynamicio/core.py [hunk 3]
```diff
@@ -64,21 +67,56 @@ def __init__(
...
+        self.schema = active_schema
+        self.name = self.schema.name.upper()
+        self.schema_validations = self.schema.validations
+        self.schema_metrics = self.schema.metrics
 
         assert hasattr(self, f"_read_from_{source_name}") or hasattr(
             self, f"_write_to_{source_name}"
         ), f"No method '_read_from_{source_name}' or '_write_to_{source_name}'. Have you registered a mixin for {source_name}?"
 
+    def _schema_from_obj(self, target) -> DataframeSchema:
```
Comment (Christos-Hadjinikolis) [inline dynamicio/core.py:89]: Make static

File: dynamicio/config/pydantic/config.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import dynamicio.config.pydantic.io_resources as env_spec
+import dynamicio.config.pydantic.table_schema as table_spec
+
+
+class IOBinding(pydantic.BaseModel):
+    """A binding for a single i/o object"""
+
+    name: str = pydantic.Field(alias="__binding_name__")
+    environments: typing.Mapping[str, env_spec.IOEnvironment]
+    dynamicio_schema: typing.Union[table_spec.DataframeSchema, table_spec.DataframeSchemaRef, None] = pydantic.Field(default=None, alias="schema")
```
Comment (Christos-Hadjinikolis) [inline dynamicio/config/pydantic/config.py:17]: Can you import the typehint you want at the start to make types more readable? 

E.g.: 

```python 
from  typing import Mapping

...
environments: Mapping[str, env_spec.IOEnvironment]
...
```

**Review by Christos-Hadjinikolis — CHANGES_REQUESTED** [review]
LGTM, though I am getting a lot of complaints from `mypy`: 

Screenshot 2022-12-27 at 12 47 28

Some of them appear to be mypy-bugs (/e/g the IOEnvironment subclasses should be recognised... ). But others appear to be legit.

**Comment by VRGhost** [comment]
mypy stuff fixed.

File: .flake8 [hunk 5]
```diff
@@ -15,5 +15,7 @@ max-line-length = 185
 max-complexity = 18
 select = B,C,E,F,W,T,I,B
 
+ban-relative-imports = true
+
```
Comment (Christos-Hadjinikolis) [inline .flake8:19]: Thanks!

File: dynamicio/config/pydantic/io_resources.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+@enum.unique
+class FileType(str, enum.Enum):
+    """List of supported file formats."""
+
+    # pylint: disable=invalid-name
+    parquet = "parquet"
+    csv = "csv"
+    json = "json"
+    hdf = "hdf"
+    txt = "txt"
```
Comment (Christos-Hadjinikolis) [inline dynamicio/config/pydantic/io_resources.py:38]: Let's remove this type here please as we don't support it.

**Review by Christos-Hadjinikolis — CHANGES_REQUESTED** [review]
Minor change requested.

**Review by Christos-Hadjinikolis — APPROVED** [review]
LGTM

**Pull request merged**